pkg/imageuploader/imageuploader.go

```diff
@@ -0,0 +1,136 @@
+package imageuploader
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/mensatt/backend/pkg/utils"
+)
+
+var (
+	isValidImageHash = regexp.MustCompile(`^[0-9a-f]{64}$`)
+
+	// https://www.garykessler.net/library/file_sigs.html
+	magicNumbers = map[string][]byte{
+		"jpg":  {0xff, 0xd8, 0xff},
+		"jpeg": {0xff, 0xd8, 0xff},
+		"png":  {0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a},
+		"webp": {0x52, 0x49, 0x46, 0x46},
+		"heif": {0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70, 0x68, 0x65, 0x69, 0x63},
+	}
+)
+
+type Config struct {
+	ImageDirectory string
+	MaxImageSizeMB int32
+	MaxResolution  int32
+}
+
+type ImageUploader struct {
+	imageDirectory string
+	maxImageSize   int64
+	maxResolution  int
+}
+
+func NewImageUploader(params Config) (*ImageUploader, error) {
+	stat, err := os.Stat(params.ImageDirectory)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("image directory does not exist - path: %s", params.ImageDirectory)
+	} else if err != nil {
+		return nil, fmt.Errorf("error checking image directory - path: %s, error: %s", params.ImageDirectory, err)
+	} else if !stat.IsDir() {
+		return nil, fmt.Errorf("image directory is not a directory - path: %s", params.ImageDirectory)
+	}
+
+	err = utils.CreateDirIfNotExists(params.ImageDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ImageUploader{
+		imageDirectory: params.ImageDirectory,
+		maxImageSize:   int64(params.MaxImageSizeMB) * 1024 * 1024,
```

Comment:
why is this an int64, it is casted to an int in the function below

Reply:
Resolution != File Size